### PR TITLE
Ericming/undertow instrumentation

### PR DESCRIPTION
**Pull request by eric-ming2**
# What Does This Do
Add instrumentation for Undertow servers

# Motivation
Adding the server instrumentation catches exceptions that happen on the server level that the servlet instrumentation does not catch.

# Additional Notes
Wildfly servers use an Undertow subsystem, so this instrumentation applies to Wildfly servers.

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -255,6 +255,7 @@ include ':dd-java-agent:instrumentation:testng-6.4'
 include ':dd-java-agent:instrumentation:tomcat-5.5'
 include ':dd-java-agent:instrumentation:trace-annotation'
 include ':dd-java-agent:instrumentation:twilio'
+include ':dd-java-agent:instrumentation:undertow-1.4'
```
Comment (charliegracie) [inline settings.gradle:258]: Does this instrumentation support Undertow 1.4+ or 2.0+? If it is 2.0+ the instrumentation directory should be changed to `undertow-2.0`.

File: dd-java-agent/instrumentation/undertow-1.4/src/main/java/datadog/trace/instrumentation/undertow/UndertowInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import static datadog.trace.bootstrap.instrumentation.api.AgentTracer.activeScope;
+import static java.util.Collections.singletonMap;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+@AutoService(Instrumenter.class)
+public final class UndertowInstrumentation extends Instrumenter.Tracing {
+
+  public UndertowInstrumentation() {
+    super("undertow", "wildfly");
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-1.4/src/main/java/datadog/trace/instrumentation/undertow/UndertowInstrumentation.java:29]:
```suggestion
    super("undertow", "undertow-1.4");
```

If you change it to `undertow-2.0` change this to match

File: dd-java-agent/instrumentation/undertow-1.4/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import static datadog.trace.instrumentation.undertow.UndertowDecorator.DECORATE;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.isPublic;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+@AutoService(Instrumenter.class)
+public final class HandlerInstrumentation extends Instrumenter.Tracing {
+
+  public HandlerInstrumentation() {
+    super("undertow", "wildfly");
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-1.4/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java:25]:
```suggestion
    super("undertow", "undertow-1.4");
```

If you change the instrumentation name to `undertow-2.0` make this the same

File: dd-java-agent/instrumentation/undertow-1.4/src/test/groovy/UndertowTest.groovy [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  @Override
+  String expectedOperationName() {
+    return 'undertow-http.request';
+  }
+
+  @Override
+  boolean testExceptionBody() {
+    false
+  }
+
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-1.4/src/test/groovy/UndertowTest.groovy:110]: The tests are not testing path parameters. Can you add the appropriate changes to add testing for path parameters. Below is an example of how to enable path parameter testing in the HttpServerTest framework. You will also need to add a new endpoint handler.

```
  @Override
  String testPathParam() {
    "/path/{id}/param"
  }
  ```

**Review by charliegracie — COMMENTED** [review]
I left some suggestions for things to change

File: settings.gradle (hunk 1 above)
Comment (eric-ming2) [reply settings.gradle]: Changed directory name to `undertow-2.0`

File: dd-java-agent/instrumentation/undertow-1.4/src/main/java/datadog/trace/instrumentation/undertow/UndertowInstrumentation.java (hunk 2 above)
Comment (eric-ming2) [reply dd-java-agent/instrumentation/undertow-1.4/src/main/java/datadog/trace/instrumentation/undertow/UndertowInstrumentation.java]: Fixed!

File: dd-java-agent/instrumentation/undertow-1.4/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java (hunk 3 above)
Comment (eric-ming2) [reply dd-java-agent/instrumentation/undertow-1.4/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java]: Fixed!

File: dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    };
+  }
+
+  // @Advice.Argument(value = 0, readOnly = false) HttpServerExchange exchange
+  public static class HandlerAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static AgentScope onEnter(
+        @Advice.Argument(value = 0, readOnly = false) HttpServerExchange exchange) {
+
+      AgentSpan existingSpan = exchange.getAttachment(DD_SPAN_ATTRIBUTE);
```
Comment (cataphract) [inline dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java:61]: The handlers can be chained, so I think this advice should be skipped in case of nested calls. `CallDepthThreadLocalMap` can help with that, something like (assuming no other paths of the before advice can return `null`):

```
@@ -58,6 +60,11 @@ public final class HandlerInstrumentation extends Instrumenter.Tracing {
     public static AgentScope onEnter(
         @Advice.Argument(value = 0, readOnly = false) HttpServerExchange exchange) {
 
+      final int callDepth = CallDepthThreadLocalMap.incrementCallDepth(HttpHandler.class);
+      if (callDepth > 0) {
+        return null;
+      }
+
       AgentSpan existingSpan = exchange.getAttachment(DD_SPAN_ATTRIBUTE);
 
       if (existingSpan != null) {
@@ -85,6 +92,11 @@ public final class HandlerInstrumentation extends Instrumenter.Tracing {
 
     @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
     public static void closeScope(@Advice.Enter final AgentScope scope) {
+      if (scope == null) {
+        return;
+      }
+      CallDepthThreadLocalMap.reset(HttpHandler.class);
+
       scope.span().finishWork();
       scope.close();
```

Comment (cataphract) [reply dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java:61]: Looking at the code, the chained handler can be synchronously if blocking mode is on or asynchronously (I imagine possibly in another thread)

**Comment by mcculls** [comment]
FYI, now https://github.com/DataDog/dd-trace-java/pull/3387 has been merged you'll need to rebase/merge in master and change the new instrumentations as follows:
* `HandlerInstrumentation` should implement `Instrumenter.ForTypeHierarchy `
* its `typeMatcher()` method should be renamed to `hierarchyMatcher`
* `UndertowInstrumentation` should implement `Instrumenter.ForSingleType `
* its `typeMatcher()` method should be removed and replaced with:
```
  @Override
  public String instrumentedType() {
    return "io.undertow.server.HttpServerExchange";
  }
```

File: dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java (hunk 5 above)
Comment (charliegracie) [reply dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java]: The code you have highlighted @cataphract was actually skipping the instrumentation after the first handler successfully handles it. As mentioned by others the call depth can not be used since handlers may be run on different threads.

File: dd-java-agent/instrumentation/micronaut-http-server-netty-2/src/main/java8/datadog/trace/instrumentation/micronaut/ChannelRead0Advice.java [hunk 6]
```diff
@@ -1,19 +1,19 @@
 package datadog.trace.instrumentation.micronaut;
 
-import static datadog.trace.bootstrap.instrumentation.api.AgentTracer.activateSpan;
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/micronaut-http-server-netty-2/src/main/java8/datadog/trace/instrumentation/micronaut/ChannelRead0Advice.java:3]: Sorry when I sent you a patch I think I accidentally included unrelated changes. You can remove the changes to the Micronaut instrumentation from PR.

Comment (eric-ming2) [reply dd-java-agent/instrumentation/micronaut-http-server-netty-2/src/main/java8/datadog/trace/instrumentation/micronaut/ChannelRead0Advice.java]: Fixed, I was responsible for that; stop trying to take credit for my mistakes.

File: dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle [hunk 7]
```diff
@@ -0,0 +1,32 @@
+muzzle {
+  pass {
+    group = "io.undertow"
+    module = "undertow-core"
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle:4]:
```suggestion
    module = "undertow-servlet"
```

Does this resolve the muzzle failure when you run locally?

Comment (eric-ming2) [reply dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle:4]: No, I get 
`Task 'muzzle-AssertPass-io.undertow-undertow-core-2.0.0.Final' not found in project ':dd-java-agent:instrumentation:undertow-2.0'.` 
for both

File: dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+apply plugin: 'org.unbroken-dome.test-sets'
+
+testSets {
+  latestDepTest {
+    dirName = 'test'
+  }
+}
+
+dependencies {
+  compileOnly group: 'io.undertow', name: 'undertow-core', version: '2.0.0.Final'
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle:21]:
```suggestion
```

File: dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  compileOnly group: 'io.undertow', name: 'undertow-servlet', version: '2.0.0.Final'
+
+  testImplementation group: 'io.undertow', name: 'undertow-core', version: '2.0.0.Final'
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle:24]:
```suggestion
```

File: dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  testImplementation group: 'io.undertow', name: 'undertow-servlet', version: '2.0.0.Final'
+
+  testImplementation project(':dd-java-agent:instrumentation:servlet')
+  testImplementation project(':dd-java-agent:instrumentation:servlet:request-3')
+
+  latestDepTestImplementation group: 'io.undertow', name: 'undertow-core', version: '2.2+'
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle:30]:
```suggestion
```

File: dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  latestDepTestImplementation group: 'io.undertow', name: 'undertow-servlet', version: '2.2+'
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle:31]:
```suggestion
  latestDepTestImplementation group: 'io.undertow', name: 'undertow-servlet', version: '+'
```

File: dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/UndertowInstrumentation.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+  @Override
+  public String[] helperClassNames() {
+    return new String[]{
+        packageName + ".ExchangeEndSpanListener",
+        packageName + ".HttpServerExchangeURIDataAdapter",
+        packageName + ".UndertowDecorator",
+        packageName + ".UndertowExtractAdapter",
+        packageName + ".UndertowExtractAdapter$Request",
+        packageName + ".UndertowExtractAdapter$Response"
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/UndertowInstrumentation.java:59]:
```suggestion
        packageName + ".UndertowExtractAdapter$Response",
        packageName + ".UndertowRunnableWrapper"
```

File: dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        packageName + ".ExchangeEndSpanListener",
+        packageName + ".HttpServerExchangeURIDataAdapter",
+        packageName + ".UndertowDecorator",
+        packageName + ".UndertowExtractAdapter",
+        packageName + ".UndertowExtractAdapter$Request",
+        packageName + ".UndertowExtractAdapter$Response"
+    };
+  }
+
+  // @Advice.Argument(value = 0, readOnly = false) HttpServerExchange exchange
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java:63]:
```suggestion
```

File: dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+  public static class HandlerAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static AgentScope onEnter(
+        @Advice.Argument(value = 0) HttpServerExchange exchange,
+        @Advice.This HttpHandler handler) {
+      // HttpHandler subclasses are chained to only the first should create a span
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java:69]:
```suggestion
      // HttpHandler subclasses are chained so only the first one should create a span
```

File: dd-java-agent/instrumentation/undertow-2.0/src/test/groovy/RedirectServlet.groovy [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+import javax.servlet.http.HttpServletResponse
+
+import static datadog.trace.agent.test.base.HttpServerTest.controller
+import static datadog.trace.agent.test.base.HttpServerTest.ServerEndpoint.REDIRECT
+
+class RedirectServlet extends HttpServlet {
+    @Override
+    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) throws ServletException, IOException {
+      controller(REDIRECT) {
+        // resp.setStatus(REDIRECT.status)
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/src/test/groovy/RedirectServlet.groovy:13]: The 3 commented lines can now be removed since sendRedirect is being used and all tests are passing

File: dd-java-agent/instrumentation/undertow-2.0/src/test/groovy/UndertowTest.groovy [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+            }
+          }
+          .addExactPath(REDIRECT.getPath()) { exchange ->
+            controller(REDIRECT) {
+              exchange.setStatusCode(StatusCodes.FOUND)
+              exchange.getResponseHeaders().put(Headers.LOCATION, REDIRECT.body)
+              exchange.endExchange()
+            }
+          }
+          // Still not sure about how to get the route properly
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/src/test/groovy/UndertowTest.groovy:62]: I would just remove this code if the path param tests can't be added at this time.

Comment (eric-ming2) [reply dd-java-agent/instrumentation/undertow-2.0/src/test/groovy/UndertowTest.groovy]: Removed

File: dd-java-agent/instrumentation/undertow-2.0/src/test/groovy/RedirectServlet.groovy (hunk 15 above)
Comment (eric-ming2) [reply dd-java-agent/instrumentation/undertow-2.0/src/test/groovy/RedirectServlet.groovy]: Removed

File: dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/UndertowInstrumentation.java (hunk 12 above)
Comment (eric-ming2) [reply dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/UndertowInstrumentation.java]: Oops, good catch!

File: dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle [hunk 17]
```diff
@@ -0,0 +1,29 @@
+muzzle {
```
Comment (charliegracie) [inline dd-java-agent/instrumentation/undertow-2.0/undertow-2.0.gradle:1]:
```suggestion
// Set properties before any plugins get loaded
ext {
  minJavaVersionForTests = JavaVersion.VERSION_1_8
}

muzzle {
```

File: dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+      DECORATE.onRequest(span, exchange, exchange, extractedContext);
+
+      final AgentScope scope = activateSpan(span);
+      scope.setAsyncPropagation(true);
+
+      // For use by servlet instrumentation
+      exchange.putAttachment(DD_UNDERTOW_SPAN, span);
+      exchange.putAttachment(DD_HTTPSERVEREXCHANGE_DISPATCH, false);
+
+      // TODO is this required?
```
Comment (eric-ming2) [inline dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/HandlerInstrumentation.java:85]: Is this required?

File: dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/UndertowRunnableWrapper.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+      DECORATE.onError(span, throwable);
+      DECORATE.onResponse(span, exchange);
+      DECORATE.beforeFinish(span);
+      span.finish();
+      throw throwable;
+    } finally {
+      if (!thrown) {
+        DECORATE.onResponse(span, exchange);
+        DECORATE.beforeFinish(span);
+        span.finish();
```
Comment (devinsba) [inline dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/UndertowRunnableWrapper.java:37]: Just a nit but this could be after super.run() in the try block so we wouldn't need a finally or the thrown variable

File: dd-java-agent/instrumentation/undertow-2.0/src/test/groovy/UndertowServletTest.groovy [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+        .setDeploymentName("servletContext.war")
+
+      builder.addServlet(new ServletInfo("SuccessServlet", SuccessServlet).addMapping(SUCCESS.getPath()))
+        .addServlet(new ServletInfo("ForwardedServlet", ForwardedServlet).addMapping(FORWARDED.getPath()))
+        .addServlet(new ServletInfo("QueryEncodedBothServlet", QueryEncodedBothServlet).addMapping(QUERY_ENCODED_BOTH.getPath()))
+        .addServlet(new ServletInfo("QueryEncodedServlet", QueryEncodedServlet).addMapping(QUERY_ENCODED_QUERY.getPath()))
+        .addServlet(new ServletInfo("QueryParamServlet", QueryServlet).addMapping(QUERY_PARAM.getPath()))
+        .addServlet(new ServletInfo("RedirectServlet", RedirectServlet).addMapping(REDIRECT.getPath()))
+        .addServlet(new ServletInfo("ErrorServlet", ErrorServlet).addMapping(ERROR.getPath()))
+        .addServlet(new ServletInfo("ExceptionServlet", ExceptionServlet).addMapping(EXCEPTION.getPath()))
```
Comment (devinsba) [inline dd-java-agent/instrumentation/undertow-2.0/src/test/groovy/UndertowServletTest.groovy:43]: If you want, you can consolidate these servlets into a single servlet like is done in `TestServlet2` and `TestServlet3`

**Review by devinsba — APPROVED** [review]

File: dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/UndertowRunnableWrapper.java (hunk 19 above)
Comment (eric-ming2) [reply dd-java-agent/instrumentation/undertow-2.0/src/main/java/datadog/trace/instrumentation/undertow/UndertowRunnableWrapper.java:37]: This is much cleaner, thanks
